server/channels/db/migrations/mysql/000107_threadmemberships_cleanup.up.sql

```diff
@@ -0,0 +1,7 @@
+DELETE FROM
+    tm USING ThreadMemberships AS tm
+    JOIN Threads ON Threads.PostId = tm.PostId
+    LEFT JOIN ChannelMembers ON ChannelMembers.UserId = tm.UserId
+    AND Threads.ChannelId = ChannelMembers.ChannelId
+WHERE
+    ChannelMembers.ChannelId IS NULL;
```

Comment:
I tried using a subquery to avoid the LEFT JOIN with ChannelMembers:
```suggestion
DELETE FROM
    tm USING ThreadMemberships AS tm
    JOIN Threads ON Threads.PostId = tm.PostId
WHERE
    (tm.UserId, Threads.ChannelId) NOT IN (SELECT UserId, ChannelId FROM ChannelMembers);
```
I tested this on MySQL 8.0.32.  The plan for origina; query:
```
mysql> explain analyze DELETE FROM
    ->     tm USING ThreadMemberships AS tm
    ->     JOIN Threads ON Threads.PostId = tm.PostId
    ->     LEFT JOIN ChannelMembers ON ChannelMembers.UserId = tm.UserId
    ->     AND Threads.ChannelId = ChannelMembers.ChannelId
    -> WHERE
    ->     ChannelMembers.ChannelId IS NULL;
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Delete from tm (buffered)  (cost=5274185.65 rows=3263648) (actual time=25909.557..25909.557 rows=0 loops=1)
    -> Filter: (ChannelMembers.ChannelId is null)  (cost=4947820.83 rows=3263648) (actual time=22.447..25803.810 rows=2766769 loops=1)
        -> Nested loop antijoin  (cost=4947820.83 rows=3263648) (actual time=22.445..25613.329 rows=2766769 loops=1)
            -> Nested loop inner join  (cost=1357807.78 rows=3263648) (actual time=1.691..12929.871 rows=3528790 loops=1)
                -> Covering index scan on Threads using idx_threads_channel_id_last_reply_at  (cost=93924.10 rows=815091) (actual time=0.884..1067.340 rows=846313 loops=1)
                -> Index lookup on tm using PRIMARY (PostId=Threads.PostId)  (cost=1.15 rows=4) (actual time=0.010..0.014 rows=4 loops=846313)
            -> Single-row covering index lookup on ChannelMembers using PRIMARY (ChannelId=Threads.ChannelId, UserId=tm.UserId)  (cost=1.00 rows=1) (actual time=0.004..0.004 rows=0 loops=3528790)
 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (25,91 sec)
```
And the plan for the modified one:
```
mysql> explain analyze DELETE FROM
    ->     tm USING ThreadMemberships AS tm
    ->     JOIN Threads ON Threads.PostId = tm.PostId
    -> WHERE
    ->     (tm.UserId, Threads.ChannelId) NOT IN (SELECT UserId, ChannelId FROM ChannelMembers);
+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| -> Delete from tm (buffered)  (cost=1619009.75 rows=3263648) (actual time=24893.593..24893.593 rows=0 loops=1)
    -> Nested loop inner join  (cost=1292644.93 rows=3263648) (actual time=1.727..24787.444 rows=2766769 loops=1)
        -> Covering index scan on Threads using idx_threads_channel_id_last_reply_at  (cost=93924.10 rows=815091) (actual time=0.088..451.006 rows=846313 loops=1)
        -> Filter: <in_optimizer>((tm.UserId,Threads.ChannelId),<exists>(select #2) is false)  (cost=1.07 rows=4) (actual time=0.014..0.028 rows=3 loops=846313)
            -> Index lookup on tm using PRIMARY (PostId=Threads.PostId)  (cost=1.07 rows=4) (actual time=0.007..0.011 rows=4 loops=846313)
            -> Select #2 (subquery in condition; dependent)
                -> Limit: 1 row(s)  (cost=0.57 rows=1) (actual time=0.004..0.004 rows=0 loops=3528790)
                    -> Filter: (<is_not_null_test>(ChannelMembers.UserId) and <if>(outer_field_is_not_null, <is_not_null_test>(ChannelMembers.ChannelId), true))  (cost=0.57 rows=1) (actual time=0.004..0.004 rows=0 loops=3528790)
                        -> Filter: ((<cache>(tm.UserId) = ChannelMembers.UserId) and <if>(outer_field_is_not_null, ((<cache>(Threads.ChannelId) = ChannelMembers.ChannelId) or <cache>((ChannelMembers.ChannelId is null))), true))  (cost=0.57 rows=1) (actual time=0.004..0.004 rows=0 loops=3528790)
                            -> Alternative plans for IN subquery: Index lookup unless ChannelId IS NULL  (cost=0.57 rows=1) (actual time=0.003..0.003 rows=0 loops=3528790)
                                -> Single-row covering index lookup on ChannelMembers using PRIMARY (ChannelId=<cache>(Threads.ChannelId), UserId=<cache>(tm.UserId))  (cost=0.57 rows=1) (actual time=0.003..0.003 rows=0 loops=3528790)
                                -> Table scan on ChannelMembers  (cost=78000.08 rows=4589138) (never executed)
 |
 ```
The plan cost is ~5 times lower for the query that avoids the LEFT JOIN. **However**, the time is very similar and none of the queries end up deleting anything in my local setup (which is based on the 12M dump, but has been modified since then). @koox00, if you have your setup still running, could you try comparing both plans? If you're using MySQL 5.7, even better, then we'll know how it works there.

Reply:
@agarciamontoro I don't have the same env ATM, but I can set it up and try again :)

Reply:
@agarciamontoro execution time for the second plan, seems better by ~20%

```sql
Query OK, 2766769 rows affected (3 min 48.43 sec)
```

And these are the EXPLAIN

```sql
mysql> EXPLAIN
    -> DELETE FROM
    ->     tm USING ThreadMemberships AS tm
    ->     JOIN Threads ON Threads.PostId = tm.PostId
    -> WHERE
    ->     (tm.UserId, Threads.ChannelId) NOT IN (SELECT UserId, ChannelId FROM ChannelMembers);
*************************** 1. row ***************************
           id: 1
  select_type: DELETE
        table: tm
   partitions: NULL
         type: ALL
possible_keys: PRIMARY
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 733432
     filtered: 100.00
        Extra: NULL
*************************** 2. row ***************************
           id: 1
  select_type: PRIMARY
        table: Threads
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 106
          ref: mattermost_test.tm.PostId
         rows: 1
     filtered: 100.00
        Extra: Using where
*************************** 3. row ***************************
           id: 2
  select_type: DEPENDENT SUBQUERY
        table: ChannelMembers
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY,idx_channelmembers_user_id_channel_id_last_viewed_at,idx_channelmembers_channel_id_scheme_guest_user_id
          key: PRIMARY
      key_len: 212
          ref: func,func
         rows: 1
     filtered: 100.00
        Extra: Using where; Using index; Full scan on NULL key
3 rows in set (0.00 sec)
```

and current one:

```sql
mysql> EXPLAIN
    -> DELETE FROM
    ->     tm USING ThreadMemberships AS tm
    ->     JOIN Threads ON Threads.PostId = tm.PostId
    ->     LEFT JOIN ChannelMembers ON ChannelMembers.UserId = tm.UserId
    ->     AND Threads.ChannelId = ChannelMembers.ChannelId
    -> WHERE
    ->     ChannelMembers.ChannelId IS NULL;
*************************** 1. row ***************************
           id: 1
  select_type: DELETE
        table: tm
   partitions: NULL
         type: ALL
possible_keys: PRIMARY
          key: NULL
      key_len: NULL
          ref: NULL
         rows: 733432
     filtered: 100.00
        Extra: NULL
*************************** 2. row ***************************
           id: 1
  select_type: SIMPLE
        table: Threads
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY
          key: PRIMARY
      key_len: 106
          ref: mattermost_test.tm.PostId
         rows: 1
     filtered: 100.00
        Extra: NULL
*************************** 3. row ***************************
           id: 1
  select_type: SIMPLE
        table: ChannelMembers
   partitions: NULL
         type: eq_ref
possible_keys: PRIMARY,idx_channelmembers_user_id_channel_id_last_viewed_at,idx_channelmembers_channel_id_scheme_guest_user_id
          key: PRIMARY
      key_len: 212
          ref: mattermost_test.Threads.ChannelId,mattermost_test.tm.UserId
         rows: 1
     filtered: 100.00
        Extra: Using where; Not exists; Using index
3 rows in set (0.00 sec)
```

Reply:
Then I think it's safe to go with the modified version for MySQL. cc/@agnivade for another pair of eyes.

Reply:
Nice, this is great stuff. This is more or less the same as what we did for another major customer. Overall, it seems like MySQL is bad at doing anti-joins, but postgres works just fine. The workaround is to explicitly write a `NOT IN` clause. Effectively both are the same really, the query optimization engine should compile them into the same logic. But we are talking about MySQL here ... :man_shrugging: 

@koox00 @agarciamontoro - This might be a great topic to bring up in the dev meeting. I think others should know about this.

Reply:
Yeah, we could make a quick presentation about it, good idea.